### DISCUSSION TITLE
Perject workflow hints/discussion

### DISCUSSION THREAD

**Discussion by shipmints** [overideal/perject#4] [General]
Greetings and thank you, again, for your (lock-down era?) work and community contribution.

With tab-bar (not tab-line), savehist, recentf, built-in project.el, and both perject-mode and perject-tab-mode enabled, I've tried creating a collection and working with projects but am having trouble grokking things. I'm sure this is all pilot error/lack of education, and I have read through some of the code to understand more.

Would you be willing to enumerate your workflow(s), which perject commands you invoke and in what order for which workflow to achieve your workflow styles? If you use any additional automation specific to the hooks you documented, e.g., using tab-group names or buffer/project associations to auto-assign frames/tabs/buffers or whitelist/blacklist buffers (exclude *earmuffs* buffers like *occur* but include *eshell* *shell*) would be helpful to understand in more detail than the documentation.

I'd be happy to share our general work-set management ideas/expectations. We've tried lots of existing packages and none, so far, yet meet our needs out of the box and I've been contributing to https://github.com/alphapapa/activities.el and am working on enhancements to https://github.com/florommel/bufferlo as I experiment with that.

-Stephane

**Comment by overideal** [comment]
Hello, please excuse my late reply.

Yeah, the project/workspace managers are always hard to grasp and adapt to one's own workflow. Since you asked for my workflow:
First of all, I use the default configuration from the README file (for perject, savehist, perject-tab, perject-consult, perject-ibuffer).

Essentially, I have an "org" collection (kind of my default "workspace") which is always open and usually a "math" collection,
containing one project for each "math-related project" i am working on.

Say I need to fix a bug in perject. I have set up a project within the collection "elisp", so I press "\<C-home\>" (`perject-open-close-or-reload`) and open the collection "elisp". This restores all elisp projects I have worked on in the recent past and creates a single frame belonging to this project. Within this "elisp" frame I select the "perject" project. Thus I can access all previous window configurations, all source files, documentation, pdfs, etc. that I may need to implement a change. I can always switch project if I want to work on something different. After I have finished the work, I close the project ("C-u \<C-home\>") and have perject delete the corresponding frame and buffers.
This leaves me with my usual "org" frame and my "math" frame (which are the two collections I have almost always open).
These are automatically restored if I exit Emacs and start it again later.

I do not have the tab-bar visible. The only way I interact with window configurations is via the `perject-tab` commands. For me, tabs (window configurations) are simply part of the project (in the perject sense) I am working on. In particular, I do not use tab-group names.

Regarding auto-assigning: This is controlled by the variable `perject-auto-add-hooks`. If you are within a frame belonging to a project, simply opening a new file (C-x C-f) that is not already open will add the file to the current project. This is a variable that advanced users may want to edit in order to increase the situations where buffers are automatically added to a project.
For example, this is one of the only perject variables I have modified in my config. Namely, I have added support for some packages I use by adding `diff-mode-hook magit-mode-hook deadgrep-mode-hook` to the list.
The goal is that you almost never have to manually call `perject-add-buffer-to-project`, because the hooks take care of all of your use cases.

To return to your question, you ask to exclude occur buffers (i.e. they should be never automatically added to a project). To achieve this, simply remove `occur-mode-hook` from the `perject-auto-add-hooks` variable. To include eshell-buffers, add `eshell-mode-hook` to the list:
```
(add-to-list 'perject-auto-add-hooks 'eshell-mode-hook)
```
Note that after modifying this particular variable, you need to disable and re-enable `perject-mode` (or restart Emacs).

Hope this helps.

**Reply by shipmints** [reply]
Thank you for the abject perject insights.
